File: engine/src/game_tree.rs
use sha2::{Digest, Sha256};

use crate::evaluator::PositionEvaluator;
use crate::math;
use crate::types::*;

/// Builds and expands game trees for the minimax search.
///
/// The game tree alternates between Agent moves (maximizing) and Adversary
/// responses (minimizing). Adversary moves model MEV bot behaviour such as
/// sandwich attacks, front-running, back-running, and JIT liquidity provision.
#[derive(Debug, Clone)]
pub struct GameTreeBuilder {
    evaluator: PositionEvaluator,
}

impl GameTreeBuilder {
    pub fn new(evaluator: PositionEvaluator) -> Self {
        Self { evaluator }
    }

    /// Build a complete game tree up to `max_depth` plies.
    ///
    /// The root node represents the current state. At even depths the Agent
    /// moves; at odd depths the Adversary responds.
    pub fn build_tree(
        &self,
        state: &OnChainState,
        actions: &[ExecutionAction],
        threats: &[MevThreat],
        max_depth: u32,
    ) -> GameNode {
        let root_hash = Self::hash_state(state);
        let mut root = GameNode::new_root(root_hash);

        if max_depth == 0 || actions.is_empty() {
            root.is_terminal = true;
            root.score = self.evaluator.evaluate_static(state);
            return root;
        }

        self.build_recursive(&mut root, state, actions, threats, max_depth, 0);
        root
    }

    /// Recursively build the tree by alternating Agent and Adversary plies.
    fn build_recursive(
        &self,
        node: &mut GameNode,
        state: &OnChainState,
        available_actions: &[ExecutionAction],
        threats: &[MevThreat],
        max_depth: u32,
        current_depth: u32,
    ) {
        if current_depth >= max_depth {
            node.is_terminal = true;
            node.score = self.evaluator.evaluate_static(state);
            return;
        }

        match node.player {
            Player::Agent => {
                // Generate agent moves
                let moves = if available_actions.is_empty() {
                    Self::generate_agent_moves(state)
                } else {
                    available_actions.to_vec()
                };

                if moves.is_empty() {
                    node.is_terminal = true;
                    node.score = self.evaluator.evaluate_static(state);
                    return;
                }

                for action in &moves {
                    let new_state = Self::simulate_action(state, action);
                    let child_hash = Self::hash_state(&new_state);
                    let mut child = GameNode::new_child(
                        action.clone(),
                        child_hash,
                        current_depth + 1,
                        Player::Adversary,
                    );

                    // Adversary responds to the agent's action
                    let adversary_threats =
                        Self::generate_adversary_moves(&new_state, action);
                    let combined_threats: Vec<MevThreat> = threats
                        .iter()
                        .chain(adversary_threats.iter())
                        .cloned()
                        .collect();

                    // Convert threats to adversary actions for the next ply
                    let threat_actions: Vec<ExecutionAction> = combined_threats
                        .iter()
                        .map(|t| Self::threat_to_action(t))
                        .collect();

                    self.build_recursive(
                        &mut child,
                        &new_state,
                        &threat_actions,
                        &combined_threats,
                        max_depth,
                        current_depth + 1,
                    );

                    node.children.push(child);
                }
            }
            Player::Adversary => {
                // Adversary ply: model each threat as a possible response
                if available_actions.is_empty() && threats.is_empty() {
                    // No adversary response: "pass" node
                    node.is_terminal = true;
                    node.score = self.evaluator.evaluate_static(state);
                    return;
                }

                // Create one child per threat / adversary action
                let adversary_actions = if !available_actions.is_empty() {
                    available_actions.to_vec()
                } else {
                    threats
                        .iter()
                        .map(|t| Self::threat_to_action(t))
                        .collect()
                };

                if adversary_actions.is_empty() {
                    node.is_terminal = true;
                    node.score = self.evaluator.evaluate_static(state);
                    return;
                }

                for adv_action in &adversary_actions {
                    let new_state = Self::simulate_action(state, adv_action);
                    let child_hash = Self::hash_state(&new_state);
                    let mut child = GameNode::new_child(
                        adv_action.clone(),
                        child_hash,
                        current_depth + 1,
                        Player::Agent,
                    );

                    // Agent gets to respond again with the original action set
                    let agent_moves = Self::generate_agent_moves(&new_state);
                    self.build_recursive(
                        &mut child,
                        &new_state,
                        &agent_moves,
                        threats,
                        max_depth,
                        current_depth + 1,
                    );

                    node.children.push(child);
                }

                // Also add a "no MEV" child where the adversary does nothing
                let pass_hash = format!("{}_pass", node.state_hash);
                let mut pass_child = GameNode {
                    action: None,
                    state_hash: pass_hash,
                    children: Vec::new(),
                    score: 0.0,
                    depth: current_depth + 1,
                    is_terminal: false,
                    player: Player::Agent,
                };
                let agent_moves = Self::generate_agent_moves(state);
                self.build_recursive(
                    &mut pass_child,
                    state,
                    &agent_moves,
                    threats,
                    max_depth,
                    current_depth + 1,
                );
                node.children.push(pass_child);
            }
        }
    }

    /// Expand a single node by generating its children.
    pub fn expand_node(&self, node: &mut GameNode, state: &OnChainState) {
        if node.is_terminal || !node.children.is_empty() {
            return;
        }

        match node.player {
            Player::Agent => {
                let moves = Self::generate_agent_moves(state);
                for action in moves {
                    let new_state = Self::simulate_action(state, &action);
                    let hash = Self::hash_state(&new_state);
                    let child =
                        GameNode::new_child(action, hash, node.depth + 1, Player::Adversary);
                    node.children.push(child);
                }
            }
            Player::Adversary => {
                // Generate adversary responses based on the node's action
                if let Some(ref action) = node.action {
                    let threats = Self::generate_adversary_moves(state, action);
                    for threat in threats {
                        let adv_action = Self::threat_to_action(&threat);
                        let new_state = Self::simulate_action(state, &adv_action);
                        let hash = Self::hash_state(&new_state);
                        let child = GameNode::new_child(
                            adv_action,
                            hash,
                            node.depth + 1,
                            Player::Agent,
                        );
                        node.children.push(child);
                    }
                }
            }
        }

        if node.children.is_empty() {
            node.is_terminal = true;
            node.score = self.evaluator.evaluate_static(state);
        }
    }

    /// Generate candidate moves for the Agent based on the current state.
    ///
    /// Inspects token balances and available pools to create feasible actions.
    pub fn generate_agent_moves(state: &OnChainState) -> Vec<ExecutionAction> {
        let mut moves = Vec::new();

        for pool in &state.pool_states {
            // Try swapping token A -> B
            if let Some(&balance_a) = state.token_balances.get(&pool.token_a_mint) {
                if balance_a > 0 && pool.reserve_a > 0 && pool.reserve_b > 0 {
                    // Swap a fraction: 10%, 25%, 50%
                    for &frac in &[10u64, 25, 50] {
                        let amount = balance_a.saturating_mul(frac) / 100;
                        if amount > 0 {
                            moves.push(ExecutionAction::new(
                                ActionKind::Swap,
                                &pool.token_a_mint,
                                amount,
                                &pool.token_b_mint,
                                50, // 0.5% default slippage tolerance
                                &pool.address,
                                5000,
                            ));
                        }
                    }
                }
            }

            // Try swapping token B -> A
            if let Some(&balance_b) = state.token_balances.get(&pool.token_b_mint) {
                if balance_b > 0 && pool.reserve_a > 0 && pool.reserve_b > 0 {
                    for &frac in &[10u64, 25, 50] {
                        let amount = balance_b.saturating_mul(frac) / 100;
                        if amount > 0 {
                            moves.push(ExecutionAction::new(
                                ActionKind::Swap,
                                &pool.token_b_mint,
                                amount,
                                &pool.token_a_mint,
                                50,
                                &pool.address,
                                5000,
                            ));
                        }
                    }
                }
            }

            // Try adding liquidity
            if let Some(&bal) = state.token_balances.get(&pool.token_a_mint) {
                if bal > 10_000 {
                    moves.push(ExecutionAction::new(
                        ActionKind::AddLiquidity,
                        &pool.token_a_mint,
                        bal / 4,
                        &pool.address,
                        100,
                        &pool.address,
                        5000,
                    ));
                }
            }
        }

File: src/engine/game-tree.ts
/**
 * MNMX Game Tree Builder
 *
 * Constructs the adversarial game tree that the minimax engine searches.
 * Each level alternates between:
 *
 *  - Agent moves:      the on-chain actions we can execute
 *  - Adversary moves:  MEV bots' probable responses (sandwiches, frontruns, …)
 *
 * The builder also provides state-simulation methods that produce new
 * OnChainState snapshots reflecting the effect of each action, without
 * touching the network.
 */

import type {
  ExecutionAction,
  GameNode,
  MevThreat,
  OnChainState,
  PoolState,
  SearchConfig,
} from '../types/index.js';
import { constantProductSwap } from '../utils/math.js';
import { hashOnChainState } from '../utils/hash.js';

// ── Builder ─────────────────────────────────────────────────────────

export class GameTreeBuilder {
  private readonly config: SearchConfig;

  constructor(config: SearchConfig) {
    this.config = config;
  }

  /**
   * Build the root of the game tree.  The root represents the current
   * on-chain state before any action is taken.
   */
  buildTree(
    state: OnChainState,
    actions: ExecutionAction[],
    adversaryActions: MevThreat[],
  ): GameNode {
    const root: GameNode = {
      action: null,
      stateHash: hashOnChainState(state),
      children: [],
      score: 0,
      depth: 0,
      isTerminal: false,
      player: 'agent',
    };

    this.expandNodeRecursive(root, state, actions, adversaryActions, 0);
    return root;
  }

  /**
   * Expand a single node by generating all legal child moves for
   * the active player.  Returns the newly created children.
   */
  expandNode(
    node: GameNode,
    state: OnChainState,
    actions: ExecutionAction[],
    adversaryActions: MevThreat[],
  ): GameNode[] {
    if (node.depth >= this.config.maxDepth) {
      node.isTerminal = true;
      return [];
    }

    if (node.player === 'agent') {
      return this.expandAgentNode(node, state, actions, adversaryActions);
    } else {
      return this.expandAdversaryNode(node, state, actions, adversaryActions);
    }
  }

  /**
   * Generate plausible adversary (MEV bot) responses to an agent action.
   */
  generateAdversaryMoves(
    state: OnChainState,
    agentAction: ExecutionAction,
  ): MevThreat[] {
    const threats: MevThreat[] = [];

    // Only certain actions invite MEV
    if (!['swap', 'provide_liquidity', 'remove_liquidity', 'liquidate'].includes(agentAction.kind)) {
      return threats;
    }

    const pool = state.poolStates.get(agentAction.pool);
    if (!pool) return threats;

    // Sandwich threat – more probable for larger trades relative to reserves
    const totalReserve = pool.reserveA + pool.reserveB;
    const tradeRatio = totalReserve > 0n
      ? Number(agentAction.amount) / Number(totalReserve)
      : 0;

    if (tradeRatio > 0.001) {
      const sandwichCost = BigInt(Math.floor(Number(agentAction.amount) * tradeRatio * 0.5));
      threats.push({
        kind: 'sandwich',
        probability: Math.min(tradeRatio * 10, 0.85),
        estimatedCost: sandwichCost,
        sourceAddress: 'SandwichBot1111111111111111111111111111111',
        relatedPool: agentAction.pool,
        description: `Sandwich attack on ${agentAction.kind} of ${agentAction.amount} via pool ${agentAction.pool.slice(0, 8)}…`,
      });
    }

    // Frontrun threat
    if (tradeRatio > 0.005) {
      const frontrunCost = BigInt(Math.floor(Number(agentAction.amount) * tradeRatio * 0.3));
      threats.push({
        kind: 'frontrun',
        probability: Math.min(tradeRatio * 6, 0.7),
        estimatedCost: frontrunCost,
        sourceAddress: 'FrontrunBot11111111111111111111111111111111',
        relatedPool: agentAction.pool,
        description: `Frontrun on ${agentAction.kind} with estimated trade-size advantage`,
      });
    }

    // JIT liquidity
    if (agentAction.kind === 'swap' && tradeRatio > 0.01) {
      threats.push({
        kind: 'jit',
        probability: Math.min(tradeRatio * 3, 0.5),
        estimatedCost: BigInt(Math.floor(Number(agentAction.amount) * 0.002)),
        sourceAddress: 'JITProvider1111111111111111111111111111111',
        relatedPool: agentAction.pool,
        description: 'JIT liquidity provision to capture swap fees',
      });
    }

    return threats;
  }

  /**
   * Simulate the effect of an agent action on the on-chain state,
   * returning a new (cloned) state with updated balances and reserves.
   */
  simulateAction(
    state: OnChainState,
    action: ExecutionAction,
  ): OnChainState {
    const next = this.cloneState(state);

    const pool = next.poolStates.get(action.pool);
    if (!pool) return next;

    switch (action.kind) {
      case 'swap':
        this.simulateSwap(next, action, pool);
        break;
      case 'transfer':
        this.simulateTransfer(next, action);
        break;
      case 'provide_liquidity':
        this.simulateAddLiquidity(next, action, pool);
        break;
      case 'remove_liquidity':
        this.simulateRemoveLiquidity(next, action, pool);
        break;
      default:
        // For other kinds, just deduct the amount from the input token
        this.deductBalance(next, action.tokenMintIn, action.amount);
        break;
    }

    next.slot += 1;
    return next;
  }

  /**
   * Simulate the effect of an MEV threat on the state.
   */
  simulateMevResponse(
    state: OnChainState,
    threat: MevThreat,
  ): OnChainState {
    const next = this.cloneState(state);
    const pool = next.poolStates.get(threat.relatedPool);
    if (!pool) return next;

    switch (threat.kind) {
      case 'sandwich': {
        // The sandwich bot's frontleg trade shifts reserves adversarially
        const shiftAmount = threat.estimatedCost / 2n;
        pool.reserveA += shiftAmount;
        if (pool.reserveB > shiftAmount) {
          pool.reserveB -= shiftAmount;
        }
        break;
      }
      case 'frontrun': {
        // Frontrunner's trade increases input reserve
        pool.reserveA += threat.estimatedCost;
        break;
      }
      case 'backrun': {
        // Backrunner captures residual price movement – minor reserve shift
        const shift = threat.estimatedCost / 4n;
        if (pool.reserveB > shift) {
          pool.reserveB -= shift;
        }
        break;
      }
      case 'jit': {
        // JIT provider adds then removes liquidity – net effect: slight fee capture
        pool.reserveA += threat.estimatedCost;
        pool.reserveB += threat.estimatedCost;
        break;
      }
    }

    next.slot += 1;
    return next;
  }

  /**
   * Hash an on-chain state for transposition-table keying.
   */
  hashState(state: OnChainState): string {
    return hashOnChainState(state);
  }

  // ── Private: Tree Expansion ─────────────────────────────────────

  private expandNodeRecursive(
    node: GameNode,
    state: OnChainState,
    actions: ExecutionAction[],
    adversaryActions: MevThreat[],
    depth: number,
  ): void {
    if (depth >= this.config.maxDepth) {
      node.isTerminal = true;
      return;
    }

    if (node.player === 'agent') {
      for (const action of actions) {
        const nextState = this.simulateAction(state, action);
        const child: GameNode = {
          action,
          stateHash: hashOnChainState(nextState),
          children: [],
          score: 0,
          depth: depth + 1,
          isTerminal: false,
          player: 'adversary',
        };
        node.children.push(child);

        // Generate adversary moves specific to this agent action
        const threats = this.generateAdversaryMoves(nextState, action);
        if (threats.length > 0 && depth + 1 < this.config.maxDepth) {
          this.expandNodeRecursive(child, nextState, actions, threats, depth + 1);
        } else {
          child.isTerminal = true;
        }
      }
    } else {
      for (const threat of adversaryActions) {
        const nextState = this.simulateMevResponse(state, threat);
        const child: GameNode = {
          action: null,
          stateHash: hashOnChainState(nextState),
          children: [],
          score: 0,
          depth: depth + 1,
          isTerminal: false,
          player: 'agent',
        };
        // Attach threat info via a synthetic action
        (child as any)._threat = threat;
        node.children.push(child);

        if (depth + 1 < this.config.maxDepth) {
          this.expandNodeRecursive(child, nextState, actions, [], depth + 1);
        } else {
          child.isTerminal = true;
        }
      }

      // Also consider "adversary does nothing" (passes)
      const passChild: GameNode = {
        action: null,
        stateHash: hashOnChainState(state),
        children: [],
        score: 0,
        depth: depth + 1,
        isTerminal: depth + 1 >= this.config.maxDepth,
        player: 'agent',
      };
      node.children.push(passChild);

      if (depth + 1 < this.config.maxDepth) {
        this.expandNodeRecursive(passChild, state, actions, [], depth + 1);
      }
    }
  }

  private expandAgentNode(
    node: GameNode,
    state: OnChainState,
    actions: ExecutionAction[],
    adversaryActions: MevThreat[],
  ): GameNode[] {
    const children: GameNode[] = [];
    for (const action of actions) {
      const nextState = this.simulateAction(state, action);
      const child: GameNode = {
        action,
        stateHash: hashOnChainState(nextState),
        children: [],
        score: 0,
        depth: node.depth + 1,
        isTerminal: node.depth + 1 >= this.config.maxDepth,
        player: 'adversary',
      };
      children.push(child);
    }
    node.children = children;
    return children;
  }

  private expandAdversaryNode(
    node: GameNode,
    state: OnChainState,
    actions: ExecutionAction[],
    adversaryActions: MevThreat[],
  ): GameNode[] {
    const children: GameNode[] = [];
    for (const threat of adversaryActions) {
      const nextState = this.simulateMevResponse(state, threat);
      const child: GameNode = {
        action: null,
        stateHash: hashOnChainState(nextState),
        children: [],
        score: 0,
        depth: node.depth + 1,
        isTerminal: node.depth + 1 >= this.config.maxDepth,
        player: 'agent',
      };
      children.push(child);
    }
    node.children = children;
    return children;
  }

  // ── Private: State Simulation ───────────────────────────────────

  private simulateSwap(
    state: OnChainState,
    action: ExecutionAction,
    pool: PoolState,
  ): void {
    const [reserveIn, reserveOut, isAtoB] = action.tokenMintIn === pool.tokenMintA
      ? [pool.reserveA, pool.reserveB, true] as const
      : [pool.reserveB, pool.reserveA, false] as const;

    const output = constantProductSwap(action.amount, reserveIn, reserveOut, pool.feeBps);

    // Update balances
    this.deductBalance(state, action.tokenMintIn, action.amount);
    this.addBalance(state, action.tokenMintOut, output);

    // Update pool reserves
    if (isAtoB) {
      pool.reserveA += action.amount;
      pool.reserveB -= output;
    } else {
      pool.reserveB += action.amount;
      pool.reserveA -= output;
    }
  }

  private simulateTransfer(
    state: OnChainState,
    action: ExecutionAction,
  ): void {
    this.deductBalance(state, action.tokenMintIn, action.amount);
  }

  private simulateAddLiquidity(
    state: OnChainState,
    action: ExecutionAction,
    pool: PoolState,
  ): void {
    const halfAmount = action.amount / 2n;
    this.deductBalance(state, action.tokenMintIn, halfAmount);
    this.deductBalance(state, action.tokenMintOut, halfAmount);
    pool.reserveA += halfAmount;
    pool.reserveB += halfAmount;
  }

  private simulateRemoveLiquidity(
    state: OnChainState,
    action: ExecutionAction,
    pool: PoolState,
  ): void {
    const halfAmount = action.amount / 2n;
    this.addBalance(state, action.tokenMintIn, halfAmount);
    this.addBalance(state, action.tokenMintOut, halfAmount);
    if (pool.reserveA > halfAmount) pool.reserveA -= halfAmount;
    if (pool.reserveB > halfAmount) pool.reserveB -= halfAmount;
  }

  // ── Private: Utility ────────────────────────────────────────────

  private cloneState(state: OnChainState): OnChainState {
    const clonedBalances = new Map(state.tokenBalances);
    const clonedPools = new Map<string, PoolState>();
    for (const [k, v] of state.poolStates) {
      clonedPools.set(k, { ...v });
    }
    return {
      tokenBalances: clonedBalances,
      poolStates: clonedPools,
      pendingTransactions: [...state.pendingTransactions],
      slot: state.slot,
      timestamp: state.timestamp,
    };
  }

  private deductBalance(
    state: OnChainState,
    mint: string,
    amount: bigint,
  ): void {
    const current = state.tokenBalances.get(mint) ?? 0n;
    state.tokenBalances.set(mint, current > amount ? current - amount : 0n);
  }

  private addBalance(
    state: OnChainState,
    mint: string,
    amount: bigint,
  ): void {
    const current = state.tokenBalances.get(mint) ?? 0n;
    state.tokenBalances.set(mint, current + amount);
  }
}
